{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eeb5c54c",
   "metadata": {},
   "source": [
    "# Miniprot mapping\n",
    "Setting up the required Python packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9237d955",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-10-11 10:33:06.607\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.config\u001b[0m:\u001b[36m<module>\u001b[0m:\u001b[36m11\u001b[0m - \u001b[1mPROJ_ROOT path is: /home/gyk/project/spider_silkome\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import time\n",
    "\n",
    "from spider_silkome_module import (\n",
    "    RAW_DATA_DIR,\n",
    "    INTERIM_DATA_DIR,\n",
    "    EXTERNAL_DATA_DIR,\n",
    "    PROCESSED_DATA_DIR\n",
    ")\n",
    "from spider_silkome_module import (\n",
    "    Attributes,\n",
    "    GFFData\n",
    ")\n",
    "from spider_silkome_module import (\n",
    "    extract_positions_from_gff,\n",
    "    positions_export\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0201bd95",
   "metadata": {},
   "source": [
    "Definition of the path for relevant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2535d457",
   "metadata": {},
   "outputs": [],
   "source": [
    "spider_species_file = f\"{EXTERNAL_DATA_DIR}/organisms.csv\"\n",
    "mechanical_properties_file = f\"{EXTERNAL_DATA_DIR}/mechanical_properties.csv\"\n",
    "spidroin_fasta_file = f\"{EXTERNAL_DATA_DIR}/spider-silkome-database.v1.prot.fixed.fasta\"\n",
    "spider_genome_path = f\"{RAW_DATA_DIR}/spider_genome\"\n",
    "spidroin_path = f\"{INTERIM_DATA_DIR}/spidroin\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98575d4b",
   "metadata": {},
   "source": [
    "Perform redundancy reduction on the N/C terminal sequences of Spidroin proteins using MMseq2.\n",
    "\n",
    "<zh>使用 MMseq2 对 Spidroin 蛋白的 N/C 端序列进行去冗余。</zh>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "96ab84e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mmseqs_output_dir=f\"{INTERIM_DATA_DIR}/mmseqs\"\n",
    "spidroin_fasta_file_rep = f\"{mmseqs_output_dir}/{spidroin_fasta_file.split('/')[-1].replace('.fasta', '')}_rep_seq.fasta\"\n",
    "if not os.path.exists(spidroin_fasta_file_rep):\n",
    "    os.makedirs(mmseqs_output_dir, exist_ok=True)\n",
    "    cmd = f\"pixi run --environment mmseqs mmseqs easy-cluster {spidroin_fasta_file} {mmseqs_output_dir}/{spidroin_fasta_file.split('/')[-1].replace('.fasta', '')} {mmseqs_output_dir}/tmp --min-seq-id 0.9 -c 0.8 --cov-mode 1\"\n",
    "    subprocess.run(cmd, shell=True)\n",
    "spidroin_fasta_file_rep_manually = spidroin_fasta_file_rep.replace('.fasta', '_manually.fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbb68ca3",
   "metadata": {},
   "source": [
    "Use miniprot to align the C-terminal and N-terminal sequences of the spidroin gene sequences to the genome of the new species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b2f321b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexing Trichonephila_clavata already exists\n",
      "Alignment Trichonephila_clavata already exists\n",
      "mRNA gff file already exists\n",
      "Total number of spidroins in Trichonephila_clavata: 18\n",
      "['MaSp1', 'MaSp3B', 'MaSp3', 'CySp', 'AcSp', 'AgSp2', 'Putative_spidroin', 'MaSp2', 'CrSp', 'MaSp', 'PySp', 'Flag', 'MiSp', 'Ampullate_spidroin', 'Pflag', 'Spidroin', 'MaSp2B', 'AgSp1']\n",
      "Indexing Araneus_ventricosus already exists\n",
      "Alignment Araneus_ventricosus already exists\n",
      "mRNA gff file already exists\n",
      "Total number of spidroins in Araneus_ventricosus: 17\n",
      "['MaSp1', 'MaSp3B', 'MaSp3', 'CySp', 'AcSp', 'AgSp2', 'Putative_spidroin', 'MaSp2', 'MaSp', 'PySp', 'Flag', 'MiSp', 'Ampullate_spidroin', 'Pflag', 'Spidroin', 'MaSp2B', 'AgSp1']\n",
      "Araneus_ventricosus.mRNA.MaSp1.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp3B.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp3.gff already exists\n",
      "Araneus_ventricosus.mRNA.CySp.gff already exists\n",
      "Araneus_ventricosus.mRNA.AcSp.gff already exists\n",
      "Araneus_ventricosus.mRNA.AgSp2.gff already exists\n",
      "Araneus_ventricosus.mRNA.Putative_spidroin.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp2.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp.gff already exists\n",
      "Araneus_ventricosus.mRNA.PySp.gff already exists\n",
      "Araneus_ventricosus.mRNA.Flag.gff already exists\n",
      "Araneus_ventricosus.mRNA.MiSp.gff already exists\n",
      "Araneus_ventricosus.mRNA.Ampullate_spidroin.gff already exists\n",
      "Araneus_ventricosus.mRNA.Pflag.gff already exists\n",
      "Araneus_ventricosus.mRNA.Spidroin.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp2B.gff already exists\n",
      "Araneus_ventricosus.mRNA.AgSp1.gff already exists\n"
     ]
    }
   ],
   "source": [
    "spider_genomes = [f for f in os.listdir(spider_genome_path) if f.endswith(\".fa.gz\")]\n",
    "gnome_mpi_path = f\"{INTERIM_DATA_DIR}/genome_mpi\"\n",
    "miniprot_output_path = f\"{INTERIM_DATA_DIR}/miniprot\"\n",
    "os.makedirs(gnome_mpi_path, exist_ok=True)\n",
    "os.makedirs(miniprot_output_path, exist_ok=True)\n",
    "for spider_genome in spider_genomes:\n",
    "    spider = spider_genome.split(\".\")[0]\n",
    "    index_cmd = f\"miniprot -t70 -d {gnome_mpi_path}/{spider}.mpi {spider_genome_path}/{spider_genome}\"\n",
    "    if not os.path.exists(f\"{gnome_mpi_path}/{spider}.mpi\"):\n",
    "        indext_start_time = time.time()\n",
    "        subprocess.run(index_cmd, shell=True)\n",
    "        indext_end_time = time.time()\n",
    "        print(f\"Indexing {spider} takes {indext_end_time - indext_start_time} seconds\")\n",
    "    else:\n",
    "        print(f\"Indexing {spider} already exists\")\n",
    "\n",
    "    output_dir = f\"{miniprot_output_path}/{spider}_all\"\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    align_cmd = f\"miniprot -t 70 -I --gff {gnome_mpi_path}/{spider}.mpi {spidroin_fasta_file_rep_manually} > {output_dir}/{spider}.gff\"\n",
    "    if not os.path.exists(f\"{output_dir}/{spider}.gff\"):\n",
    "        align_start_time = time.time()\n",
    "        subprocess.run(align_cmd, shell=True)\n",
    "        align_end_time = time.time()\n",
    "        print(f\"Alignment {spider} takes {align_end_time - align_start_time} seconds\")\n",
    "    else:\n",
    "        print(f\"Alignment {spider} already exists\")\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{spider}.mRNA.gff\"):\n",
    "        grep_cmd = f\"grep 'mRNA' {output_dir}/{spider}.gff | sort -k1,1V -k4,4n > {output_dir}/{spider}.mRNA.gff\"\n",
    "        subprocess.run(grep_cmd, shell=True)\n",
    "    else:\n",
    "        print(f\"mRNA gff file already exists\")\n",
    "\n",
    "    if os.path.exists(f\"{output_dir}/{spider}.mRNA.gff\"):\n",
    "        mRNA_gff = pd.read_csv(f\"{output_dir}/{spider}.mRNA.gff\", sep='\\t', header=None)\n",
    "        gff_header = [\"seqid\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "        mRNA_gff.columns = gff_header\n",
    "        spidroins = list(set([row[\"attribute\"].split(';')[-1].split('|')[-2] for index, row in mRNA_gff.iterrows()]))\n",
    "        print(f\"Total number of spidroins in {spider}: {len(spidroins)}\\n{spidroins}\")\n",
    "        for spidroin in spidroins:\n",
    "            gff_spidroin_output = f\"{output_dir}/{spider}.mRNA.{spidroin}.gff\"\n",
    "            if not os.path.exists(gff_spidroin_output):\n",
    "                grep_cmd = f\"grep '|{spidroin}|' {output_dir}/{spider}.mRNA.gff > {gff_spidroin_output}\"\n",
    "                subprocess.run(grep_cmd, shell=True)\n",
    "            else:\n",
    "                print(f\"{spider}.mRNA.{spidroin}.gff already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cf09a51",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "\n",
    "spider = \"Trichonephila_clavata\"\n",
    "output_dir = f\"{miniprot_output_path}/{spider}_all\"\n",
    "mRNA_gff = f\"{output_dir}/{spider}.mRNA.gff\"\n",
    "\n",
    "merged_mRNA_gff = BedTool(mRNA_gff).merge(s=True)\n",
    "for gene in merged_mRNA_gff:\n",
    "    print(gene)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "735c8192",
   "metadata": {},
   "source": [
    "Defining the gene boundaries for each type of Spidroin should be part of the `for spider_genome in spider_genomes:` loop. For convenience, this section has been separated for testing.\n",
    "\n",
    "<zh>处理每一种蛛丝蛋白的基因边界界定，这一部分应该作为`for spider_genome in spider_genomes:`循环下的一部分，这里为了方便，独立出来进行测试。</zh>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb17eb51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-10-11 10:40:14.608\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.608\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.608\u001b[0m | \u001b[33m\u001b[1mWARNING \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m186\u001b[0m - \u001b[33m\u001b[1mNo data to export\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.610\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.610\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.612\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Pflag.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.612\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.612\u001b[0m | \u001b[33m\u001b[1mWARNING \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m186\u001b[0m - \u001b[33m\u001b[1mNo data to export\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.613\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Pflag.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.613\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.701\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MiSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.701\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 54\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.702\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 3\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.702\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 51\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.702\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MiSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.703\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.712\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.713\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 51\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.713\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.713\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 47\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.714\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.714\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.720\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AcSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.721\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.721\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.721\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 3\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.722\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AcSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.722\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 3\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.728\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CySp.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.728\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.728\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.728\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 3\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.729\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CySp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.729\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 2\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.733\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3B.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.733\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 16\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.733\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.733\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 12\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.734\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3B.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.734\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.741\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Spidroin.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.742\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 42\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.742\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 3\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.742\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 39\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.743\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Spidroin.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.743\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 12\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.744\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Putative_spidroin.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.744\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.744\u001b[0m | \u001b[33m\u001b[1mWARNING \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m186\u001b[0m - \u001b[33m\u001b[1mNo data to export\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.745\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Putative_spidroin.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.745\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.746\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CrSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.746\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.746\u001b[0m | \u001b[33m\u001b[1mWARNING \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m186\u001b[0m - \u001b[33m\u001b[1mNo data to export\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.746\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CrSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.747\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.753\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp2.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.753\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 7\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.753\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.753\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 6\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.754\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp2.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.754\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.767\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp1.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.768\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 53\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.768\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 4\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.768\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 49\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.769\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp1.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.769\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 5\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.777\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Ampullate_spidroin.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.777\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 7\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.777\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.777\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 6\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.778\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Ampullate_spidroin.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.778\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 5\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.788\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.788\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 30\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.789\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 2\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.789\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 28\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.789\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.790\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 2\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.798\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Flag.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.798\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 5\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.798\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 2\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.799\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 3\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.799\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Flag.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.799\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 5\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.805\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp1.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.805\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 2\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.805\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.806\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.806\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp1.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.806\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.811\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2B.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.811\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.811\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.811\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.812\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2B.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.812\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.819\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m180\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.PySp.csv\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.819\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mTotal combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.819\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m183\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.819\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[1mInvalid combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.820\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m326\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.PySp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-11 10:40:14.820\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m327\u001b[0m - \u001b[1mTotal genes predicted: 1\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "spidroins = ['MaSp3', 'Pflag', 'MiSp', 'MaSp2', 'AcSp', 'CySp',\n",
    "'MaSp3B', 'Spidroin', 'Putative_spidroin', 'CrSp', 'AgSp2', 'MaSp1',\n",
    "'Ampullate_spidroin', 'MaSp', 'Flag', 'AgSp1', 'MaSp2B', 'PySp'] # Define the spidroins for test, should be the same as the spidroins used in the `for spider_genome in spider_genomes:` loop.\n",
    "all_gff_records = []\n",
    "all_csv = pd.DataFrame()\n",
    "for spidroin in spidroins:\n",
    "    # spidroin = \"MiSp\"\n",
    "    spider = \"Trichonephila_clavata\"\n",
    "    output_dir = f\"{miniprot_output_path}/{spider}_all\"\n",
    "    spidroin_gff = f\"{output_dir}/{spider}.mRNA.{spidroin}.gff\"\n",
    "    attr_dict = {}\n",
    "    spidroin_gff_data = []\n",
    "    with open(spidroin_gff, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "\n",
    "            fields = line.strip().split('\\t')\n",
    "\n",
    "            # Analysis of the attributes field\n",
    "            for attr in fields[8].split(';'):\n",
    "                if '=' in attr:\n",
    "                    key, value = attr.split('=', 1)\n",
    "                    attr_dict[key] = value\n",
    "\n",
    "            # Create attributes object\n",
    "            attr_obj = Attributes(\n",
    "                ID=attr_dict['ID'],\n",
    "                Rank=int(attr_dict['Rank']),\n",
    "                Identity=float(attr_dict['Identity']),\n",
    "                Positive=float(attr_dict['Positive']),\n",
    "                Target=attr_dict['Target'].split('|')\n",
    "            )\n",
    "\n",
    "            # Create gff_data object and add to list\n",
    "            spidroin_gff_data.append(GFFData(\n",
    "                seqid=fields[0],\n",
    "                source=fields[1],\n",
    "                type=fields[2],\n",
    "                start=int(fields[3]),\n",
    "                end=int(fields[4]),\n",
    "                score=float(fields[5]),\n",
    "                strand=fields[6],\n",
    "                frame=fields[7],\n",
    "                attributes=attr_obj\n",
    "            ))\n",
    "\n",
    "    # Sort by positive with descending order\n",
    "    spidroin_gff_data.sort(key=lambda x: x.attributes.Positive, reverse=True)\n",
    "    # Extract positions\n",
    "    positions = extract_positions_from_gff(spidroin_gff_data, positive_threshold=0.75)\n",
    "    # Export data\n",
    "    # Export CSV data\n",
    "    csv_output = spidroin_gff.replace('.gff', '.csv')\n",
    "    df = positions_export(positions, csv_output, format='csv')\n",
    "    all_csv = pd.concat([all_csv, df], ignore_index=True)\n",
    "\n",
    "    # Export GFF data\n",
    "    gff_output = spidroin_gff.replace('.gff', '.combined.gff')\n",
    "    gff_records = positions_export(\n",
    "        positions,\n",
    "        gff_output,\n",
    "        format='gff',\n",
    "        spidroin=spidroin,\n",
    "        min_length=1000, # Set minimum gene length threshold\n",
    "        max_length=100000, # Set maximum gene length threshold\n",
    "        extension_length=10000 # Set length to extend when start or end is missing\n",
    ")\n",
    "\n",
    "    all_gff_records.extend(gff_records)\n",
    "\n",
    "# Convert to DataFrame and save\n",
    "final_output_dir = f\"{PROCESSED_DATA_DIR}/01.miniprot_mapping\"\n",
    "os.makedirs(final_output_dir, exist_ok=True)\n",
    "if all_gff_records:\n",
    "    df_combined = pd.DataFrame(all_gff_records)\n",
    "    # Sort by seqid (chromosome) and start position with natural sorting\n",
    "    df_combined['seqid_sort'] = df_combined['seqid'].str.extract(r'(\\d+)').astype(float)\n",
    "    df_combined = df_combined.sort_values(['seqid_sort', 'start']).drop('seqid_sort', axis=1)\n",
    "    # //todo: 添加一个函数，能够处理 df_comined 中基因座相似的情况。\n",
    "    # Write to GFF file\n",
    "    with open(f\"{final_output_dir}/{spider}.gff\", 'w') as f:\n",
    "        f.write(\"##gff-version 3\\n\")\n",
    "        df_combined.to_csv(f, sep='\\t', header=False, index=False)\n",
    "else:\n",
    "    print(f\"No GFF records to combine for {spider}\")\n",
    "\n",
    "if not all_csv.empty:\n",
    "    all_csv.to_csv(f\"{final_output_dir}/{spider}.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
