{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eeb5c54c",
   "metadata": {},
   "source": [
    "# Spidroin curation\n",
    "Setting up the required Python packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9237d955",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-10-10 13:48:28.875\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.config\u001b[0m:\u001b[36m<module>\u001b[0m:\u001b[36m11\u001b[0m - \u001b[1mPROJ_ROOT path is: /home/gyk/project/spider_silkome\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import time\n",
    "\n",
    "from spider_silkome_module import (\n",
    "    RAW_DATA_DIR,\n",
    "    INTERIM_DATA_DIR,\n",
    "    EXTERNAL_DATA_DIR,\n",
    "    PROCESSED_DATA_DIR\n",
    ")\n",
    "from spider_silkome_module import (\n",
    "    Attributes,\n",
    "    GFFData\n",
    ")\n",
    "from spider_silkome_module import (\n",
    "    extract_positions_from_gff,\n",
    "    positions_export\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0201bd95",
   "metadata": {},
   "source": [
    "Definition of the path for relevant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2535d457",
   "metadata": {},
   "outputs": [],
   "source": [
    "spider_species_file = f\"{EXTERNAL_DATA_DIR}/organisms.csv\"\n",
    "mechanical_properties_file = f\"{EXTERNAL_DATA_DIR}/mechanical_properties.csv\"\n",
    "spidroin_fasta_file = f\"{EXTERNAL_DATA_DIR}/spider-silkome-database.v1.prot.fixed.fasta\"\n",
    "spider_genome_path = f\"{RAW_DATA_DIR}/spider_genome\"\n",
    "spidroin_path = f\"{INTERIM_DATA_DIR}/spidroin\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98575d4b",
   "metadata": {},
   "source": [
    "使用 MMseq2 对 Spidroin 蛋白的 N/C 端序列进行去冗余。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "96ab84e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster /home/gyk/project/spider_silkome/data/external/spider-silkome-database.v1.prot.fixed.fasta /home/gyk/project/spider_silkome/data/interim/mmseqs/spider-silkome-database.v1.prot.fixed /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp --min-seq-id 0.9 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t18.8cc5c\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Compositional bias scale            \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Mask lower letter repeating N times \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t80\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.9\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "Use GPU                             \t0\n",
      "\n",
      "createdb /home/gyk/project/spider_silkome/data/external/spider-silkome-database.v1.prot.fixed.fasta /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input --createdb-mode 1 --write-lookup 0 \n",
      "\n",
      "Shuffle database cannot be combined with --createdb-mode 1\n",
      "We recompute with --shuffle 0\n",
      "Converting sequences\n",
      "[Multiline fasta can not be combined with --createdb-mode 0\n",
      "We recompute with --createdb-mode 1\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "[=\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 1ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 89ms\n",
      "Create directory /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp\n",
      "cluster /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp --max-seqs 20 -c 0.8 --cov-mode 1 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.9 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 1\n",
      "linclust /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 80 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 80 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 80 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 11155 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 11.15K 0s 12ms\n",
      "Sort kmer 0h 0m 0s 7ms\n",
      "Sort by rep. sequence 0h 0m 0s 5ms\n",
      "Time for fill: 0h 0m 0s 5ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 64ms\n",
      "rescorediagonal /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 11.15K 0s 10ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 9ms\n",
      "Time for processing: 0h 0m 0s 64ms\n",
      "clust /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_rescore1 /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 80 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 13ms\n",
      "\n",
      "Size of the sequence database: 11155\n",
      "Size of the alignment database: 11155\n",
      "Number of clusters: 8481\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 20ms\n",
      "createsubdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/order_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "createsubdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/order_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "filterdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_filter1 /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_filter2 --filter-file /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/order_redundancy --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 8.48K 0s 8ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 47ms\n",
      "rescorediagonal /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_filter2 /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 8.48K 0s 11ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 68ms\n",
      "align /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_rescore2 /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 8481 type: Aminoacid\n",
      "Target database size: 8481 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 8.48K 0s 153ms\n",
      "Time for merging to aln: 0h 0m 0s 8ms\n",
      "18607 alignments calculated\n",
      "9660 sequence pairs passed the thresholds (0.519159 of overall calculated)\n",
      "1.139017 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 204ms\n",
      "clust /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/aln /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 80 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 12ms\n",
      "\n",
      "Size of the sequence database: 8481\n",
      "Size of the alignment database: 8481\n",
      "Number of clusters: 7530\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "mergeclusters /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pre_clust /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/clust --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 8.48K 0s 7ms\n",
      "Clustering step 2\n",
      "[=================================================================] 7.53K 0s 18ms\n",
      "Write merged clustering\n",
      "[=================================================================] 11.15K 0s 51ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 68ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/linclust/1171994881477103379/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "prefilter /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 7530 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 7530 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 7.53K 0s 21ms\n",
      "Index table: Masked residues: 1326835\n",
      "Index table: fill\n",
      "[=================================================================] 7.53K 0s 15ms\n",
      "Index statistics\n",
      "Entries:          461555\n",
      "DB size:          490 MB\n",
      "Avg k-mer size:   0.007212\n",
      "Top 10 k-mers\n",
      "    NMFSAE\t1520\n",
      "    PGSCLV\t881\n",
      "    CDLQLE\t681\n",
      "    QTGVQF\t597\n",
      "    SKKQNM\t538\n",
      "    NERLMF\t498\n",
      "    QFNILI\t480\n",
      "    ALMFSM\t451\n",
      "    CELQLE\t423\n",
      "    ASPLCD\t382\n",
      "Time for index table init: 0h 0m 0s 683ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 7530\n",
      "Target db start 1 to 7530\n",
      "[=================================================================] 7.53K 0s 32ms\n",
      "\n",
      "0.602821 k-mers per position\n",
      "3330 DB matches per sequence\n",
      "0 overflows\n",
      "18 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 924ms\n",
      "align /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/pref_step0 /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 7530 type: Aminoacid\n",
      "Target database size: 7530 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 7.53K 1s 685ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 7ms\n",
      "93405 alignments calculated\n",
      "8352 sequence pairs passed the thresholds (0.089417 of overall calculated)\n",
      "1.109163 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 733ms\n",
      "clust /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/aln_step0 /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 80 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 11ms\n",
      "\n",
      "Size of the sequence database: 7530\n",
      "Size of the alignment database: 7530\n",
      "Number of clusters: 7136\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "mergeclusters /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_redundancy /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_step0 --threads 80 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 7.53K 0s 6ms\n",
      "Clustering step 2\n",
      "[=================================================================] 7.14K 0s 19ms\n",
      "Write merged clustering\n",
      "[=================================================================] 11.15K 0s 49ms\n",
      "Time for merging to clu: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 0s 65ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_tmp/14512155558865545692/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/cluster.tsv --threads 80 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 52ms\n",
      "result2repseq /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_rep --db-load-mode 0 --compressed 0 --threads 80 -v 3 \n",
      "\n",
      "[=================================================================] 7.14K 0s 5ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 17ms\n",
      "Time for processing: 0h 0m 0s 75ms\n",
      "result2flat /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_rep /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "createseqfiledb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_seqs --threads 80 -v 3 \n",
      "\n",
      "[=================================================================] 7.14K 0s 9ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 0s 48ms\n",
      "result2flat /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_seqs /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 28ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb /home/gyk/project/spider_silkome/data/interim/mmseqs/tmp/18438224705311594293/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n"
     ]
    }
   ],
   "source": [
    "mmseqs_output_dir=f\"{INTERIM_DATA_DIR}/mmseqs\"\n",
    "spidroin_fasta_file_rep = f\"{mmseqs_output_dir}/{spidroin_fasta_file.split('/')[-1].replace('.fasta', '')}_rep_seq.fasta\"\n",
    "if not os.path.exists(spidroin_fasta_file_rep):\n",
    "    os.makedirs(mmseqs_output_dir, exist_ok=True)\n",
    "    cmd = f\"pixi run --environment mmseqs mmseqs easy-cluster {spidroin_fasta_file} {mmseqs_output_dir}/{spidroin_fasta_file.split('/')[-1].replace('.fasta', '')} {mmseqs_output_dir}/tmp --min-seq-id 0.9 -c 0.8 --cov-mode 1\"\n",
    "    subprocess.run(cmd, shell=True)\n",
    "spidroin_fasta_file_rep_manually = spidroin_fasta_file_rep.replace('.fasta', '_manually.fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbb68ca3",
   "metadata": {},
   "source": [
    "Use miniprot to align the C-terminal and N-terminal sequences of the spidroin gene sequences to the genome of the new species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2b2f321b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexing Trichonephila_clavata already exists\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::mp_idx_restore@3.536*1.00] loaded the index\n",
      "[M::mp_mapopt_set_max_intron] set max intron size to 184590\n",
      "[M::mp_idx_print_stat] 2364518 distinct k-mers; mean occ of infrequent k-mers: 391.92; 573 frequent k-mers accounting for 15240950 occurrences\n",
      "[M::worker_pipeline::26.333*54.50] mapped 4096 sequences\n",
      "[M::worker_pipeline::43.620*54.62] mapped 3040 sequences\n",
      "[M::main] Version: 0.18-r281\n",
      "[M::main] CMD: miniprot -t 70 -I --gff /home/gyk/project/spider_silkome/data/interim/genome_mpi/Trichonephila_clavata.mpi /home/gyk/project/spider_silkome/data/interim/mmseqs/spider-silkome-database.v1.prot.fixed_rep_seq_manually.fasta\n",
      "[M::main] Real time: 43.951 sec; CPU: 2382.903 sec; Peak RSS: 17.852 GB\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment Trichonephila_clavata takes 44.06230330467224 seconds\n",
      "Total number of spidroins in Trichonephila_clavata: 18\n",
      "['MaSp3', 'Pflag', 'MiSp', 'MaSp2', 'AcSp', 'CySp', 'MaSp3B', 'Spidroin', 'Putative_spidroin', 'CrSp', 'AgSp2', 'MaSp1', 'Ampullate_spidroin', 'MaSp', 'Flag', 'AgSp1', 'MaSp2B', 'PySp']\n",
      "Indexing Araneus_ventricosus already exists\n",
      "Alignment Araneus_ventricosus already exists\n",
      "mRNA gff file already exists\n",
      "Total number of spidroins in Araneus_ventricosus: 17\n",
      "['Pflag', 'MaSp3', 'MiSp', 'CySp', 'AcSp', 'MaSp3B', 'Putative_spidroin', 'Spidroin', 'AgSp2', 'MaSp1', 'Ampullate_spidroin', 'MaSp', 'Flag', 'AgSp1', 'MaSp2B', 'MaSp2', 'PySp']\n",
      "Araneus_ventricosus.mRNA.Pflag.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp3.gff already exists\n",
      "Araneus_ventricosus.mRNA.MiSp.gff already exists\n",
      "Araneus_ventricosus.mRNA.CySp.gff already exists\n",
      "Araneus_ventricosus.mRNA.AcSp.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp3B.gff already exists\n",
      "Araneus_ventricosus.mRNA.Putative_spidroin.gff already exists\n",
      "Araneus_ventricosus.mRNA.Spidroin.gff already exists\n",
      "Araneus_ventricosus.mRNA.AgSp2.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp1.gff already exists\n",
      "Araneus_ventricosus.mRNA.Ampullate_spidroin.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp.gff already exists\n",
      "Araneus_ventricosus.mRNA.Flag.gff already exists\n",
      "Araneus_ventricosus.mRNA.AgSp1.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp2B.gff already exists\n",
      "Araneus_ventricosus.mRNA.MaSp2.gff already exists\n",
      "Araneus_ventricosus.mRNA.PySp.gff already exists\n"
     ]
    }
   ],
   "source": [
    "spidroin_files = os.listdir(spidroin_path)\n",
    "spider_genomes = [f for f in os.listdir(spider_genome_path) if f.endswith(\".fa.gz\")]\n",
    "gnome_mpi_path = f\"{INTERIM_DATA_DIR}/genome_mpi\"\n",
    "miniprot_output_path = f\"{INTERIM_DATA_DIR}/miniprot\"\n",
    "os.makedirs(gnome_mpi_path, exist_ok=True)\n",
    "os.makedirs(miniprot_output_path, exist_ok=True)\n",
    "for spider_genome in spider_genomes:\n",
    "    spider = spider_genome.split(\".\")[0]\n",
    "    index_cmd = f\"miniprot -t70 -d {gnome_mpi_path}/{spider}.mpi {spider_genome_path}/{spider_genome}\"\n",
    "    if not os.path.exists(f\"{gnome_mpi_path}/{spider}.mpi\"):\n",
    "        indext_start_time = time.time()\n",
    "        subprocess.run(index_cmd, shell=True)\n",
    "        indext_end_time = time.time()\n",
    "        print(f\"Indexing {spider} takes {indext_end_time - indext_start_time} seconds\")\n",
    "    else:\n",
    "        print(f\"Indexing {spider} already exists\")\n",
    "\n",
    "    output_dir = f\"{miniprot_output_path}/{spider}_all\"\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    align_cmd = f\"miniprot -t 70 -I --gff {gnome_mpi_path}/{spider}.mpi {spidroin_fasta_file_rep_manually} > {output_dir}/{spider}.gff\"\n",
    "    if not os.path.exists(f\"{output_dir}/{spider}.gff\"):\n",
    "        align_start_time = time.time()\n",
    "        subprocess.run(align_cmd, shell=True)\n",
    "        align_end_time = time.time()\n",
    "        print(f\"Alignment {spider} takes {align_end_time - align_start_time} seconds\")\n",
    "    else:\n",
    "        print(f\"Alignment {spider} already exists\")\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{spider}.mRNA.gff\"):\n",
    "        grep_cmd = f\"grep 'mRNA' {output_dir}/{spider}.gff > {output_dir}/{spider}.mRNA.gff\"\n",
    "        subprocess.run(grep_cmd, shell=True)\n",
    "    else:\n",
    "        print(f\"mRNA gff file already exists\")\n",
    "\n",
    "    if os.path.exists(f\"{output_dir}/{spider}.mRNA.gff\"):\n",
    "        mRNA_gff = pd.read_csv(f\"{output_dir}/{spider}.mRNA.gff\", sep='\\t', header=None)\n",
    "        gff_header = [\"seqid\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "        mRNA_gff.columns = gff_header\n",
    "        spidroins = list(set([row[\"attribute\"].split(';')[-1].split('|')[-2] for index, row in mRNA_gff.iterrows()]))\n",
    "        print(f\"Total number of spidroins in {spider}: {len(spidroins)}\\n{spidroins}\")\n",
    "        for spidroin in spidroins:\n",
    "            gff_spidroin_output = f\"{output_dir}/{spider}.mRNA.{spidroin}.gff\"\n",
    "            if not os.path.exists(gff_spidroin_output):\n",
    "                grep_cmd = f\"grep '|{spidroin}|' {output_dir}/{spider}.mRNA.gff > {gff_spidroin_output}\"\n",
    "                subprocess.run(grep_cmd, shell=True)\n",
    "            else:\n",
    "                print(f\"{spider}.mRNA.{spidroin}.gff already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb17eb51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2025-10-10 14:50:27.427\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.427\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 174\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.428\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 8\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.428\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 166\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.430\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.430\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 8\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.436\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Pflag.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.436\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.437\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.437\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.438\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Pflag.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.439\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.509\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MiSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.509\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 4282\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.509\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 10\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.509\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 4272\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.521\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MiSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.521\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 10\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.534\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.534\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 393\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.535\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 5\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.535\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 388\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.537\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.537\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 5\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.544\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AcSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.544\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 63\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.544\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.545\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 62\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.545\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AcSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.546\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.557\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CySp.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.557\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 528\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.558\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.558\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 528\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.561\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CySp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.561\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.565\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3B.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.565\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 17\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.565\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 4\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.565\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 13\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.566\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp3B.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.566\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 4\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.641\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Spidroin.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.641\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 290\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.641\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 16\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.641\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 274\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.643\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Spidroin.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.644\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 16\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.645\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Putative_spidroin.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.645\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.645\u001b[0m | \u001b[33m\u001b[1mWARNING \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m184\u001b[0m - \u001b[33m\u001b[1mNo data to export\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.646\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Putative_spidroin.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.646\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.649\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CrSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.649\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.649\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.649\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.650\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.CrSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.650\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 0\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.657\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp2.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.657\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 8\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.657\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.657\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 7\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.658\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp2.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.658\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.676\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp1.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.676\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 365\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.676\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 11\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.676\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 354\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.678\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp1.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.678\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 11\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.690\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Ampullate_spidroin.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.690\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 365\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.690\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 10\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.690\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 355\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.692\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Ampullate_spidroin.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.692\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 10\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.709\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.709\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 695\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.710\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 14\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.710\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 681\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.712\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.712\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 14\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.722\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Flag.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.722\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 80\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.722\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 4\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.723\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 76\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.724\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.Flag.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.724\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 4\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.731\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp1.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.731\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 36\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.731\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.731\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 35\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.732\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.AgSp1.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.732\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 1\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.736\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2B.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.736\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 54\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.736\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 3\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.737\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 51\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.738\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.MaSp2B.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.738\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 3\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.748\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m178\u001b[0m - \u001b[1mCSV saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.PySp.csv\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.748\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m179\u001b[0m - \u001b[1mTotal combinations: 302\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.748\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m181\u001b[0m - \u001b[1mValid combinations: 2\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.748\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_csv\u001b[0m:\u001b[36m182\u001b[0m - \u001b[1mInvalid combinations: 300\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.750\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m267\u001b[0m - \u001b[1mGFF saved to /home/gyk/project/spider_silkome/data/interim/miniprot/Trichonephila_clavata_all/Trichonephila_clavata.mRNA.PySp.combined.gff\u001b[0m\n",
      "\u001b[32m2025-10-10 14:50:27.750\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mspider_silkome_module.export\u001b[0m:\u001b[36m_export_to_gff\u001b[0m:\u001b[36m268\u001b[0m - \u001b[1mTotal genes predicted: 2\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "spidroins = ['MaSp3', 'Pflag', 'MiSp', 'MaSp2', 'AcSp', 'CySp',\n",
    "'MaSp3B', 'Spidroin', 'Putative_spidroin', 'CrSp', 'AgSp2', 'MaSp1',\n",
    "'Ampullate_spidroin', 'MaSp', 'Flag', 'AgSp1', 'MaSp2B', 'PySp']\n",
    "all_gff_records = []\n",
    "all_csv = pd.DataFrame()\n",
    "for spidroin in spidroins:\n",
    "    # spidroin = \"MiSp\"\n",
    "    spider = \"Trichonephila_clavata\"\n",
    "    output_dir = f\"{miniprot_output_path}/{spider}_all\"\n",
    "    spidroin_gff = f\"{output_dir}/{spider}.mRNA.{spidroin}.gff\"\n",
    "    attr_dict = {}\n",
    "    spidroin_gff_data = []\n",
    "    with open(spidroin_gff, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "\n",
    "            fields = line.strip().split('\\t')\n",
    "\n",
    "            # Analysis of the attributes field\n",
    "            for attr in fields[8].split(';'):\n",
    "                if '=' in attr:\n",
    "                    key, value = attr.split('=', 1)\n",
    "                    attr_dict[key] = value\n",
    "\n",
    "            # Create attributes object\n",
    "            attr_obj = Attributes(\n",
    "                ID=attr_dict['ID'],\n",
    "                Rank=int(attr_dict['Rank']),\n",
    "                Identity=float(attr_dict['Identity']),\n",
    "                Positive=float(attr_dict['Positive']),\n",
    "                Target=attr_dict['Target'].split('|')\n",
    "            )\n",
    "\n",
    "            # Create gff_data object and add to list\n",
    "            spidroin_gff_data.append(GFFData(\n",
    "                seqid=fields[0],\n",
    "                source=fields[1],\n",
    "                type=fields[2],\n",
    "                start=int(fields[3]),\n",
    "                end=int(fields[4]),\n",
    "                score=float(fields[5]),\n",
    "                strand=fields[6],\n",
    "                frame=fields[7],\n",
    "                attributes=attr_obj\n",
    "            ))\n",
    "\n",
    "    # Sort by positive with descending order\n",
    "    spidroin_gff_data.sort(key=lambda x: x.attributes.Positive, reverse=True)\n",
    "    # Extract positions\n",
    "    positions = extract_positions_from_gff(spidroin_gff_data, positive_threshold=0.5)\n",
    "    # Export data\n",
    "    # Export CSV data\n",
    "    csv_output = spidroin_gff.replace('.gff', '.csv')\n",
    "    df = positions_export(positions, csv_output, format='csv')\n",
    "    all_csv = pd.concat([all_csv, df], ignore_index=True)\n",
    "\n",
    "    # Export GFF data\n",
    "    gff_output = spidroin_gff.replace('.gff', '.combined.gff')\n",
    "    gff_records = positions_export(\n",
    "        positions,\n",
    "        gff_output,\n",
    "        format='gff',\n",
    "        spidroin=spidroin,\n",
    "        min_length=1000, # Set minimum gene length threshold\n",
    "        max_length=100000, # Set maximum gene length threshold\n",
    "        extension_length=10000 # Set length to extend when start or end is missing\n",
    ")\n",
    "\n",
    "    all_gff_records.extend(gff_records)\n",
    "\n",
    "# Convert to DataFrame and save\n",
    "if all_gff_records:\n",
    "    df_combined = pd.DataFrame(all_gff_records)\n",
    "    # Sort by seqid (chromosome) and start position with natural sorting\n",
    "    df_combined['seqid_sort'] = df_combined['seqid'].str.extract(r'(\\d+)').astype(float)\n",
    "    df_combined = df_combined.sort_values(['seqid_sort', 'start']).drop('seqid_sort', axis=1)\n",
    "    # Write to GFF file\n",
    "    with open(f\"{PROCESSED_DATA_DIR}/{spider}_combined.gff\", 'w') as f:\n",
    "        f.write(\"##gff-version 3\\n\")\n",
    "        df_combined.to_csv(f, sep='\\t', header=False, index=False)\n",
    "else:\n",
    "    print(f\"No GFF records to combine for {spider}\")\n",
    "\n",
    "if not all_csv.empty:\n",
    "    all_csv.to_csv(f\"{PROCESSED_DATA_DIR}/{spider}_combined.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
